### PR TITLE
[PR] Add auth

### LINKED ISSUE #1: Write a Tutorial From First Principals using Phoenix LiveView to Sync Cursors on a Canvas/Page
One of the "killer" features of `Pheonix` and now `LiveView` is the low cost websockets 
for realtime communication between _many_ connected browsers. 

# Todo

+ [x] Write an example/tutorial from scratch that showcases cursors on a page.
+ [x] Integrate with `auth` / `auth_plug` to display names and avatars (profile pictures) next to the cursor
+ [ ] Deploy to Fly.io (new org) so the demo is permanently up.

For context on _why_ we have so many separate repos to demonstrate tiny experiments, please read: https://github.com/dwyl/technology-stack/issues/96

### PR DESCRIPTION
**Pull request by LuchoTurtle**
closes #1 

This PR adds authentication using https://github.com/dwyl/auth_plug.
Instead of usernames being randomly generated and assigned, when a user logs in, his username is shown.

### CODE REVIEW

**Comment by LuchoTurtle** [comment]
After logging in, I can't seem to get access to the socket assigns pertaining to the logged-in user. 
No changes to the assigns seem to occur after logging in, contrary to the https://github.com/dwyl/auth_plug docs.

Will further look into this tomorrow, during/after standup.

**Review by Arhell — COMMENTED** [review]
cc @nelsonic

**Comment by LuchoTurtle** [comment]
Gonna have a look at this https://github.com/dwyl/mvp/blob/main/BUILDIT.md#62-create-auth-controller
this should fix whatever problems I'm having.

**Comment by LuchoTurtle** [comment]
@nelsonic the auth tests are failing because there's no `AUTH_API_KEY`. 
Is there an org one I should use?

The PR is ready and the auth works properly now.

**Comment by nelsonic** [comment]
We don't have any org-wide **`AUTH_API_KEY`** that would be a security anti-pattern. 
[don't worry if you didn't know... but it would be like using the same internet password on all websites ...]
We create a new "App" for each instance of `auth_plug` with a descriptive name 
and then have an **`AUTH_API_KEY`** that can be easily reset/revoked for that App. 
See: https://github.com/dwyl/phoenix-liveview-realtime-cursor-tracking-tutorial/issues/1#issuecomment-1325725280

File: README.md [hunk 1]
```diff
@@ -831,6 +831,334 @@ to the `assets/css/app.css` file:
 }
 ```
 
+## 5. Adding authentication 🔐
```
Comment (nelsonic) [inline README.md:834]: @LuchoTurtle please start adding whitespace between a section heading and its' content
the way _all_ our repos do, e.g: https://github.com/dwyl/learn-tdd/blame/main/README.md#L17

Yes, it's technically "valid" Markdown to _not_ have whitespace between the header and the content ...
but it doesn't _look_ right. 

Think of it like this, a "Mankini" is _technically_ a valid form of swimwear ... 🩱 
image

But it doesn't _look_ right, does it? 😬 

```suggestion
## 5. Adding authentication 🔐

```

File: README.md [hunk 2]
```diff
@@ -831,6 +831,334 @@ to the `assets/css/app.css` file:
...
+a given provider (such as Google or Github)
+and show the related username for everyone to see!
+
+For this, we can leverage the 
+[`auth_plug`](https://github.com/dwyl/auth_plug)
+package to easily integrate this feature.
+
+Let's do this!
+
+### 5a. Add a login button
```
Comment (nelsonic) [inline README.md:848]:
```suggestion
### 5a. Add a login button

```

File: README.md [hunk 3]
```diff
@@ -831,6 +831,334 @@ to the `assets/css/app.css` file:
...
+</div>
+```
+
+We've simply added a button that,
+once clicked, creates a `login` event.
+We are going to eventually handle this event,
+but before that,
+we need to add `auth_plug` to our project.
+
+### 5b. Add `auth_plug`
```
Comment (nelsonic) [inline README.md:899]:
```suggestion
### 5b. Add `auth_plug`

```

File: README.md [hunk 4]
```diff
@@ -831,6 +831,334 @@ to the `assets/css/app.css` file:
...
+    get "/logout", AuthController, :logout
+  end
+```
+
+And you should be sorted!
+If you had any trouble setting `auth_plug` running,
+[check their docs](https://github.com/dwyl/phoenix-chat-example/blob/add-auth/auth.md),
+as the guide is more thorough there.
+
+### 5c. Showing username
```
Comment (nelsonic) [inline README.md:1015]:
```suggestion
### 5c. Showing username

```

File: README.md [hunk 5]
```diff
@@ -831,6 +831,338 @@ to the `assets/css/app.css` file:
...
+
+Create a file `lib/chat_web/controllers/auth_controller.ex`
+and add the following.
+
+```elixir
+defmodule LiveCursorsWeb.AuthController do
+  use LiveCursorsWeb, :controller
+  import Phoenix.LiveView, only: [assign_new: 3]
+
+  def add_assigns(:default, _params, %{"jwt" => jwt} = _session, socket) do
```
Comment (nelsonic) [inline README.md:952]: This code should be DRYed: https://github.com/dwyl/auth_plug/issues/86 
Copy-pasting it from the MVP is a clear sign that it should be included in the library. 💭

File: README.md (hunk 1 above)
Comment (LuchoTurtle) [reply README.md:834]: You're correct, I should follow the convention used across dwyl's repos. 
But personally, to me, it looks "right" without whitespace.

It's just a matter of preference :)

File: README.md (hunk 5 above)
Comment (LuchoTurtle) [reply README.md:952]: Created this PR -> https://github.com/dwyl/auth_plug/pull/88. 
After merging and publishing a new `auth-plug` version, I can make the changes and deploy the "Live Cursors" with auth version to fly.io.

**Comment by nelsonic** [comment]
Pending review/merge of: https://github.com/dwyl/auth_plug/pull/88

File: README.md (hunk 5 above)
Comment (nelsonic) [reply README.md:952]: As of `auth_plug@1.5.1` you can now call: 

```elixir
socket = socket
  |> AuthPlug.assign_jwt_to_socket(&Phoenix.LiveView.assign_new/3, jwt)
```

Comment (nelsonic) [reply README.md:952]: Or ... in-line the whole call as:

```elixir
def add_assigns(:default, _params, %{"jwt" => jwt} = _session, socket) do
  {:cont,  
    AuthPlug.assign_jwt_to_socket(socket, &Phoenix.LiveView.assign_new/3, jwt)}
end
```

**Comment by LuchoTurtle** [comment]
Made the changes to use `AuthPlug.assign_jwt_to_socket/3`, as discussed. Also formatted the files, in case CI complains about it.

File: README.md (hunk 5 above)
Comment (nelsonic) [reply README.md:952]: The instance in the `code` is fixed by https://github.com/dwyl/phoenix-liveview-realtime-cursor-tracking-tutorial/commit/a75ef36eeb16c0b3918b01f3a2a9d4635ba9475b but not the instructions (`code` block) in the `README.md` ... 💭

Comment (LuchoTurtle) [reply README.md:952]: I'm dumb.
Fixed with https://github.com/dwyl/phoenix-liveview-realtime-cursor-tracking-tutorial/pull/12/commits/84f1687d2e74b43072c3fff7485293c17da83382

sorry

Comment (nelsonic) [reply README.md:957]: Not "dumb" at all. We just have the code in two places in these tutorials ... 🙄 
If there was a way to DRY this e.g. using `LiveBook` for all our tutorials/examples: https://github.com/dwyl/learn-phoenix-framework/issues/147 📖 
Then we could avoid the duplication. 💭

File: README.md [hunk 6]
```diff
@@ -30,13 +30,17 @@ Learn how to create a **live cursor tracking app** using **Phoenix LiveView**!
       - [Installing TailwindCSS](#installing-tailwindcss)
     - [Setup watcher, minification and loading Tailwind](#setup-watcher-minification-and-loading-tailwind)
     - [1b. Running the Phoenix app](#1b-running-the-phoenix-app)
-  - [2. LiveView with cursors :small_red_triangle:](#2-liveview-with-cursors-small_red_triangle)
+  - [2. LiveView with cursors :small\_red\_triangle:](#2-liveview-with-cursors-small_red_triangle)
```
Comment (nelsonic) [inline README.md:33]:
```suggestion
  - [2. LiveView with cursors 🔺🔺small](#2-liveview-with-cursors-small_red_triangle)
```

**Review by nelsonic — APPROVED** [review]
@LuchoTurtle looks great! Thanks. 👌

**Pull request merged**